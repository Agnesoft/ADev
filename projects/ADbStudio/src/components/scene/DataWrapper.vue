<template>
    <FadeTransition :opened="show">  
        <div class="data-wrapper"  :style="dataStyle"  @mouseenter="exploreData(true)" @mouseleave="exploreData(false)">
            <BaseBox  v-for="element in elements" :key="element.id" :def_opened="boxOpened" :closable="true">
                <h4 slot="title">{{ getTitle(element) }}</h4>
                <ElementData slot="content" v-if="dataValid(element)" :data="element.data" />
                <div slot="content" v-else class="no-data">
                    {{ $t("no-data-selected") }}
                </div>
            </BaseBox>
            
        </div>
    </FadeTransition>
</template>

<script>
import ElementData from './ElementData';
import FadeTransition from "@/components/transitions/FadeTransition";

export default {
    name: "DataWrapper",
    components: {
        ElementData,
        FadeTransition,
    },
    props: {
        elements: Array,
        show: {
            type: Boolean,
            default: true,
        },
        dataStyle: {
            type: Object,
            default() {return {}},
        }
    },
    computed:{
        boxOpened: function(){
            return this.elements.length === 1;
        }
    },
    methods: {
        dataValid(element){
            if(Object.prototype.hasOwnProperty.call(element, 'data') && Object.keys(element.data).length != 0){
                return true;
            }
            return false;
        },
        getTitle(element){
            if(element.id < 0){
                return this.$t('edge-id',{'id': element.id});
            } else {
                return this.$t('node-id',{'id': element.id});
            }
        },
        exploreData(explore){
            this.$emit('exploreData',explore);
        },
    }

}
</script>

<style scoped>
    .data-wrapper{
        background: var(--light-color);
        max-width: 0.7vw;
        min-width: 12rem;
        max-height: 30rem;
        z-index: 1000;
        box-shadow:  0 0 3px var(--primary-color), 2px 2px 5px var(--dark-color);
        overflow: auto;
    }

    .no-data{
        padding: 0.4rem;
    }
</style>
