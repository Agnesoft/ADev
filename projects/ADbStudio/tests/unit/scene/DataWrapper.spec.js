import { mount  } from '@vue/test-utils'
import DataWrapper from '@/components/scene/DataWrapper.vue'

import { localVue,i18n } from '../TestUtils'

describe('DataWrapper',() => {
    let wrapper;
    let fadeTransitionStub; 
    let boxStub; 

    beforeEach(() => {
        
        fadeTransitionStub = () => ({
            props: {
                opened: Boolean,
            },
            template: '<div v-if="opened" ><slot></slot></div>',
          });
        boxStub = () => ({
              props: {
                  opened: Boolean,
              },
              template: '<div class="base-box"> <slot name="title" /><slot name="content"></slot></div>',
            });
        wrapper = mount(DataWrapper, {
            localVue,
            i18n,
            propsData: {
                elements: [
                    { id: 1 },
                    { id: 2, data: {}},
                    { id: 3, data: {foo: "bar"}},
                ]
            },
            stubs: {
                FadeTransition:  fadeTransitionStub(),
                BaseBox: boxStub(),
            },
          });
    })
    it("render 3 boxes", () => {
        expect(wrapper.findAll(".base-box").length).toBe(3);
    });
    it("renders 'no data' in the first two boxes", () => {
        let boxes = wrapper.findAll(".base-box");
        console.log(boxes.at)
        expect(boxes.at(0).find(".no-data").exists()).toBeTruthy();
        expect(boxes.at(0).find(".element-data").exists()).toBeFalsy();
        expect(boxes.at(1).find(".no-data").exists()).toBeTruthy();
        expect(boxes.at(1).find(".element-data").exists()).toBeFalsy();
    });
    it("renders element data in the third box",async () => {
        let boxes = wrapper.findAll(".base-box");
        expect(boxes.at(2).find(".no-data").exists()).toBeFalsy();
        expect(boxes.at(2).find(".element-data").exists()).toBeTruthy();
    });
    it('renders correctly (snapshot)', () => {
        expect(wrapper.element).toMatchSnapshot()
    })
})
