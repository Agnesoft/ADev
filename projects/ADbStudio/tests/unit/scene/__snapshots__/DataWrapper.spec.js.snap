// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataWrapper renders correctly (snapshot) 1`] = `
<div>
  <div
    class="data-wrapper"
  >
    <div
      class="base-box"
      closable="true"
    >
      <h4>
        Node 1
      </h4>
      <div
        class="no-data"
      >
        
                No data selected
            
      </div>
    </div>
    <div
      class="base-box"
      closable="true"
    >
      <h4>
        Node 2
      </h4>
      <div
        class="no-data"
      >
        
                No data selected
            
      </div>
    </div>
    <div
      class="base-box"
      closable="true"
    >
      <h4>
        Node 3
      </h4>
      <table
        class="element-data"
      >
        <tr>
          <td>
            foo
          </td>
           
          <td>
            bar
          </td>
        </tr>
      </table>
    </div>
  </div>
</div>
`;
